test.sh

```diff
@@ -1,40 +1,74 @@
-#!/bin/bash
+#!/usr/bin/env bash
+
+# -e Stops execution in the instance of a command or pipeline error
+# -u Treat unset variables as an error and exit immediately
+set -eu
+
 # Run all tests and coverage checks. Called from Travis automatically, also
 # suitable to run manually. See list of prerequisite packages in .travis.yml
 if type realpath >/dev/null 2>&1 ; then
-  cd $(realpath $(dirname $0))
+  cd "$(realpath -- $(dirname -- "$0"))"
 fi
 
-set -exo pipefail
-TRAVIS=${TRAVIS:-false}
+#
+# Defaults
+#
+STATUS="FAILURE"
+# RUN, UNIT_FILTER, and INT_FILTER vars are here for backwards compatibility,
+# and should be set to () unconditionally in the future.
+if [ -z "${RUN+x}" ]; then RUN=(); fi
+if [ -z "${UNIT_FILTER+x}" ]; then UNIT_FILTER=(); fi
+if [ -z "${INT_FILTER+x}" ]; then INT_FILTER=(); fi
+ENABLE_UNIT_RACE="false"
+export BOULDER_CONFIG_DIR="${BOULDER_CONFIG_DIR:-"test/config"}"
 
-# The list of segments to run. To run only some of these segments, pre-set the
-# RUN variable with the ones you want (see .travis.yml for an example).
-# Order doesn't matter. Note: gomod-vendor is specifically left out of the
-# defaults, because we don't want to run it locally (it could delete local
-# state) We also omit coverage by default on local runs because it generates
-# artifacts on disk that aren't needed.
-RUN=${RUN:-lints unit integration}
+#
+# Print Functions
+#
+function print_outcome() {
+  if [ "$STATUS" == SUCCESS ]
+  then
+    echo -e "\e[32m"$STATUS"\e[0m"
+  else
+    echo -e "\e[31m"$STATUS"\e[0m"
+  fi
+}
+
+function print_list_of_integration_tests() {
+  for file in ./test/integration/*.go; do
+    [ -e "$file" ] || continue
+    cat "$file" | grep -e '^func Test' | awk '{print $2}' | sed s/\(t//
```

Comment:
You can do all this in a single `sed` invocation:

```
sed -n 's/.*\(Test\w*\).*/\1/p < "${file}"
```

Reply:
From my perspective, this is harder read and reason about than the line that it replaces. There are some other issues that I should point out as well:

In it's current state, it doesn't produce an equivalent output to the line it seeks to replace:
```
$ sed -n 's/.*\(Test\w*\).*/\1/p' < test/integration/akamai_purger_drain_queue_test.go
TestAkamaiPurgerDrainQueueFails
TestAkamaiPurgerDrainQueueSucceeds
TestSrvCmd
TestSrvCmd
TestSrvCmd
TestSrvCmd
TestSrvCmd
TestSrvCmd
```

```
$ cat test/integration/akamai_purger_drain_queue_test.go | grep -e '^func Test' | awk '{print $2}' | sed s/\(t//
TestAkamaiPurgerDrainQueueFails
TestAkamaiPurgerDrainQueueSucceeds
```

Unfortunately, `sed` is differently on different platforms. For instance, running the suggested line under BSD `sed` gives us:
```
$ sed -n 's/.*\(Test\w*\).*/\1/p' < test/integration/akamai_purger_drain_queue_test.go
Test
Test
Test
Test
Test
Test
Test
Test
```
Where as, running the line it seeks to replace produces:

```
$ cat test/integration/akamai_purger_drain_queue_test.go | grep -e '^func Test' | awk '{print $2}' | sed s/\(t//
TestAkamaiPurgerDrainQueueFails
TestAkamaiPurgerDrainQueueSucceeds
```

I worked very hard to keep the commands in this script portable so that we wouldn't have to worry about platform conflicts like the one above.

Reply:
Interesting, I didn't know that about BSD `sed`! Mainly I found it awkward to use both `awk` and `sed`, each a fully-featured string processing language on its own. How about this, which lets us skip the `for` loop too:

```
go test -tags integration -list=. ./test/integration/... | grep '^Test'
```

Reply:
This is perfect, thank you.